src/GPUActivityAgent.cpp

```diff
@@ -0,0 +1,464 @@
+                                    GEOPM_DOMAIN_GPU,
+                                    domain_idx), NAN});
+        }
+
+        m_time = {m_platform_io.push_signal("TIME", GEOPM_DOMAIN_BOARD, 0), NAN};
+
+        for (int domain_idx = 0; domain_idx < M_NUM_GPU; ++domain_idx) {
+            m_gpu_freq_control.push_back(control{m_platform_io.push_control("GPU_FREQUENCY_CONTROL",
+                                         GEOPM_DOMAIN_GPU,
+                                         domain_idx), NAN});
+        }
+
+        auto all_names = m_platform_io.control_names();
+        if (all_names.find("DCGM::FIELD_UPDATE_RATE") != all_names.end()) {
+            // DCGM documentation indicates that users should query no faster than 100ms
+            // even though the interface allows for setting the polling rate in the us range.
+            // In practice reducing below the 100ms value has proven functional, but should only
+            // be attempted if there is a proven need to catch short phase behavior that cannot
+            // be accomplished with the default settings.
+            m_platform_io.write_control("DCGM::FIELD_UPDATE_RATE", GEOPM_DOMAIN_BOARD, 0, 0.1); //100ms
+            m_platform_io.write_control("DCGM::MAX_STORAGE_TIME", GEOPM_DOMAIN_BOARD, 0, 1);
+            m_platform_io.write_control("DCGM::MAX_SAMPLES", GEOPM_DOMAIN_BOARD, 0, 100);
+        }
+    }
+
+    // Validate incoming policy and configure default policy requests.
+    void GPUActivityAgent::validate_policy(std::vector<double> &in_policy) const
+    {
+        assert(in_policy.size() == M_NUM_POLICY);
+        double gpu_min_freq = m_platform_io.read_signal("GPU_FREQUENCY_MIN_AVAIL", GEOPM_DOMAIN_BOARD, 0);
+        double gpu_max_freq = m_platform_io.read_signal("GPU_FREQUENCY_MAX_AVAIL", GEOPM_DOMAIN_BOARD, 0);
+
+        // Check for NAN to set default values for policy
+        if (std::isnan(in_policy[M_POLICY_GPU_FREQ_MAX])) {
+            in_policy[M_POLICY_GPU_FREQ_MAX] = gpu_max_freq;
+        }
+
+        if (in_policy[M_POLICY_GPU_FREQ_MAX] > gpu_max_freq ||
+            in_policy[M_POLICY_GPU_FREQ_MAX] < gpu_min_freq ) {
+            throw Exception("GPUActivityAgent::" + std::string(__func__) +
+                            "(): GPU_FREQ_MAX out of range: " +
+                            std::to_string(in_policy[M_POLICY_GPU_FREQ_MAX]) +
+                            ".", GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        // Not all gpus provide an 'efficient' frequency signal, and the
+        // value provided by the policy may not be valid.  In this case approximating
+        // f_efficient as midway between F_min and F_max is reasonable.
+        if (std::isnan(in_policy[M_POLICY_GPU_FREQ_EFFICIENT])) {
+            in_policy[M_POLICY_GPU_FREQ_EFFICIENT] = (in_policy[M_POLICY_GPU_FREQ_MAX]
+                                                             + gpu_min_freq) / 2;
```

Comment:
```suggestion
            in_policy[M_POLICY_GPU_FREQ_EFFICIENT] = (in_policy[M_POLICY_GPU_FREQ_MAX] + gpu_min_freq) / 2;
```

or align it one space in from the open paren on the line above:
```suggestion
            in_policy[M_POLICY_GPU_FREQ_EFFICIENT] = (in_policy[M_POLICY_GPU_FREQ_MAX]
                                                      + gpu_min_freq) / 2;
```